crates/paralegal-spdg/src/lib.rs

```diff
@@ -656,4 +677,38 @@ impl Ctrl {
             m.insert(from.into_owned(), iter::once(to).collect());
         }
     }
+
+    /// Gather all [`DataSource`]s that are mentioned in this controller including data and control flow.
+    pub fn all_sources(&self) -> HashSet<&DataSource> {
+        self.data_flow
+            .0
+            .keys()
+            .chain(self.types.0.keys())
+            .chain(self.ctrl_flow.0.keys())
+            .dedup()
+            .collect()
+    }
+
+    /// Gather all [`DataSink`]s or [`CallSite`]s that are mentioned in this controller including data and control flow.
+    pub fn all_call_sites_or_sinks(&self) -> impl Iterator<Item = CallSiteOrDataSink> + '_ {
+        self.data_flow
+            .0
+            .values()
+            .flatten()
+            .map(|s| CallSiteOrDataSink::DataSink(s.clone()))
```

Comment:
You could use `flat_map` here like this

```rs
.flat_map(|s| {
    let as_or = CallSiteOrDataSink::DataSink(s.clone());
    match s {
        DataSink::Argument { function, .. } => vec![CallSiteOrDataSink::CallSite(*function), as_or],
        _ => vec![as_or]
    }
})
```

That way you avoid the first `chain`

Reply:
Done